truffle/src/main/java/org/jruby/truffle/Main.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2015 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.jruby.truffle;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.jruby.Ruby;
+import org.jruby.RubyInstanceConfig;
+import org.jruby.ast.RootNode;
+import org.jruby.internal.runtime.GlobalVariable;
+import org.jruby.internal.runtime.ValueAccessor;
+import org.jruby.runtime.IAccessor;
+import org.jruby.runtime.ThreadContext;
+import org.jruby.runtime.builtin.IRubyObject;
+
+public class Main {
+
+    public static void main(String[] args) {
+        final RubyInstanceConfig config = new RubyInstanceConfig();
+        config.setHardExit(true);
+        config.processArguments(args);
+
+        InputStream in = config.getScriptSource();
+        String filename = config.displayedFileName();
+
+        final Ruby runtime = Ruby.newInstance(config);
+        final AtomicBoolean didTeardown = new AtomicBoolean();
+
+        config.setCompileMode(RubyInstanceConfig.CompileMode.TRUFFLE);
+
+        if (config.isHardExit()) {
+            // We're the command-line JRuby, and should set a shutdown hook for
+            // teardown.
+            Runtime.getRuntime().addShutdownHook(new Thread() {
+                public void run() {
+                    if (didTeardown.compareAndSet(false, true)) {
+                        runtime.tearDown();
+                    }
+                }
+            });
+        }
+
+        if (in == null) {
+            return;
+        } else {
+
+            // Global variables
+            IAccessor d = new ValueAccessor(runtime.newString(filename));
+            runtime.getGlobalVariables().define("$PROGRAM_NAME", d,
+                    GlobalVariable.Scope.GLOBAL);
```

Comment:
Slightly long lines are OK if it makes the call simpler to read (like in this case)